networkd: allow to supply own unit files

Mic92 opened the pull request:
###### Motivation for this change

A lot of effort was put into this module, but it only support half the options
supported by latest systemd-networkd. 
The patch allows to generate/write own files to this directory. Instead of providing /etc/systemd/network only single files are symlinked to this directory.

The following example:

```
systemd.network= {
  enable = true;
  networks.eth0 = {
    enable = true;
    name = "eth0";
    address = [ "2a03:4000:13:31e::/64" ];
    gateway = [ "fe80::1" ];
    DHCP = "v4";
    networkConfig.IPv6AcceptRA = "no";
  };
};
```

can be then also written as:

```
systemd.network.enable = true;
environment.etc."systemd/network/eth0.network".text = ''
  [Match]
  Name = eth0

  [Network]
  DHCP = v4
  Address = 2a03:4000:13:31e::/64
  Gateway = fe80::1
  IPv6AcceptRA = no
'';
```

The latter one is much easier to interfere by looking at the manpage or
copy'n'paste from other examples. Also systemd-networkd provides better
error messages compared to the module. 

In future wireguard vpn will be supported by networkd.
To avoid leakage of private keys, those could be then also put
outside of the /nix/store

Currently files changed outside of this module will not trigger update
of networkd. My suggestion therefor is to refactor the module to
just take strings instead of attributes.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Mic92 committed:
networkd: allow to supply own unit files

Networkd units can contain secrets. In future also wireguard vpn will be supported by
networkd. To avoid leakage of private keys, those could be then also put outside
of the /nix/store

Having a writeable /etc/systemd/network also allows to quick fix network issues,
when upgrading `nixos-rebuild switch` would require network on its own (due
updates).

ericsagnes commented:
What about an in-between solution, making `systemd.network` also accepting strings with a `either` so the following can be done:

```
systemd.network = {
  enable = true;
  networks.eth0 = ''
    [Match]
    Name = eth0

    [Network]
    DHCP = v4
    Address = 2a03:4000:13:31e::/64
    Gateway = fe80::1
    IPv6AcceptRA = no
  '';
};
```

or with a dedicated `configFile` option like sometime seen in other modules:

```
systemd.network = {
  enable = true;
  networks.eth0.configFile = ''
    [Match]
    Name = eth0

    [Network]
    DHCP = v4
    Address = 2a03:4000:13:31e::/64
    Gateway = fe80::1
    IPv6AcceptRA = no
  '';
};
```

So the user can chose to use either the attrs way or the copy-paste way, and the config files can still be managed by the networkd module?

Mic92 commented:
This would work in theory, but i fear that this module is currently not maintained. 
It should at least a field where keys can be passed, which are not type checked.

Mic92 committed:
networkd: add extraConfig to all units

networkd options are always correct or up to date. This option allows to by
pass type checking. It is also easier to write because examples can be just copy
and paste from manpages.

Mic92 commented:
I added extraConfig for all sections now.
